### PR TITLE
Added Cosine Distance as Metric

### PR DESCRIPTION
**Pull request by brunoalano**
I believe that using the Cosine distance can be important for some applications. Based on my tests, clustering word embeddings or sentence embeddings using Euclidean distance gives bad results. Using the Cosine Distance, improve them. I believe that's because the Norm.

### CODE REVIEW

**Comment by lmcinnes** [comment]
I agree some form of angular distance might be useful. Cosine distance has the issue that it is not actually a metric. Since most space trees, including the kd-tree and ball-tree used in hdbscan, rely on the triangle inequality (which cosine distance violates) I'm wary of using it as it may introduce some rather irregular results. You may want to consider angular distance instead (arccos(similarity)/pi) as I believe that is a metric at least -- not sure how well it will work in this case, but I'd be happy to include it if you feel that it works for you.

**Comment by brunoalano** [comment]
@lmcinnes I never heard about the arccos-distance, the only reference that I found is [1] and [2]. As usual, I always used the Cosine Similarity to this task, and as I've seen running HDBSCAN, improved the results. If you want, I can implement the Angular Distance, but I don't know if will improve.

But I can say that the Cosine Similarity it's working, I'm using a Neural Net Autoencoder to output the sequence "thought" vector (embedding).

[1] http://math.stackexchange.com/questions/102924/cosine-similarity-distance-and-triangle-equation
[2] https://en.wikipedia.org/wiki/Angular_distance

**Comment by lmcinnes** [comment]
I would be interested to know (seeing as you have a working use case) if
the arccos-distance actually performs similarly to cosine distance. I worry
that for certain code paths the cosine distance can break badly (as it
violates some fundamental assumptions). If arccos-distance is clearly worse
then we can use to cosine distance, but it would be useful to flag a
warning to the user when they select it just in case it breaks. If the
arccos-distance performs simialrly or well enough I would much prefer to
use that. As the stack-exchange link notes you'll have to be careful with
zero vectors (I would suggest they should be maximally far from everything).

**Comment by lmcinnes** [comment]
I should also say: thanks for doing this! I left out cosine distance
because, as I said it can break things in some cases. I didn't really have
any suitable data to really test with either, so I never really got around
to trying out the various options (angular distance/arccos-distance being
the prime candidate). Any work that helps extend hdbscan to make it more
useful to others is always greatly appreciated!

**Comment by brunoalano** [comment]
I will be testing now

**Comment by brunoalano** [comment]
@lmcinnes It's possible to use Cosine Distance on Sklearn `pairwise_distance` with the `ArccosDistance`?

**Comment by lmcinnes** [comment]
Yes, the pairwise_distance code path doesn't make use of trees, so ypu
should be okay to just use cosine distance there. Good catch.

**Comment by brunoalano** [comment]
I should remove the Cosine Distance now? Or should put a warning?

**Comment by lmcinnes** [comment]
I'm okay with either -- if you're happy with the arccos distance then you
can just keep that and remove cosine distance. If, on the other hand, you
still think cosine distance can be useful then by all means keep it and add
a warning that it may fail and suggest using the arccos distance.

**Comment by lmcinnes** [comment]
This was looking really promising. Had you made a decision on what to do with the cosine distance? Remove it or have a warning? I'd be really happy to see this wrapped up and rolled in!

**Comment by brunoalano** [comment]
@lmcinnes I've been out those days, but I tested both metrics and the results are pretty similar on my task. I believe that only the `arccos` distance should remain, since the cosine similarity can give errors to 0-vectors.

**Comment by brunoalano** [comment]
Other thing, I think that will be great to use the alias named `cosine` to `arccos` distance, since I think people coming from other areas don't know what mean the `arccos` distance metric. What do you think?

**Comment by lmcinnes** [comment]
I agree, sounds good! Thanks for this work.

**Comment by lmcinnes** [comment]
I put in some work recently, so I'll need to resolve some conflicts, which may take a little while, but I'll tyry to get this merged as soon as I can. Thanks again.

**Comment by brunoalano** [comment]
I hope that I can help in anything more in HDBSCAN

**Comment by lmcinnes** [comment]
Well, if you see anything more that you think would be useful, please feel free to add it, or fix it. I'm more than happy to accept further pull requests.

**Comment by lmcinnes** [comment]
That may just be the state of the data. If you can demonstrate that they
should be clustering I might be concerned, but without any evidence to the
contrary it is hard to know whether it is simply the case that the data
does not cluster well.

**Comment by brunoalano** [comment]
Where can I contact you? I can share the data with you

**Comment by lmcinnes** [comment]
You can reach me at leland.mcinnes at gmail.com
